smst_proofs_test.go

```diff
@@ -174,3 +174,129 @@ func TestSMST_ProofsSanityCheck(t *testing.T) {
+	path := sha256.Sum256([]byte("jackfruit"))
+	flipPathBit(path[:], 245)
+	closestKey, closestValueHash, closestSum, proof, err = smst.ProveClosest(path[:])
+	require.NoError(t, err)
+	require.NotEqual(t, proof, &SparseMerkleProof{})
+
+	result = VerifySumProof(proof, root, closestKey, closestValueHash, closestSum, NoPrehashSpec(sha256.New(), true))
+	require.True(t, result)
+	closestPath := sha256.Sum256([]byte("jackfruit"))
+	require.Equal(t, closestPath[:], closestKey)
+	require.Equal(t, closestSum, uint64(7))
+
+	path = sha256.Sum256([]byte("xwordA188wordB110"))
+	flipPathBit(path[:], 245)
+	closestKey, closestValueHash, closestSum, proof, err = smst.ProveClosest(path[:])
+	require.NoError(t, err)
+	require.NotEqual(t, proof, &SparseMerkleProof{})
+
+	result = VerifySumProof(proof, root, closestKey, closestValueHash, closestSum, NoPrehashSpec(sha256.New(), true))
+	require.True(t, result)
+	closestPath = sha256.Sum256([]byte("xwordA188wordB110"))
+	require.Equal(t, closestPath[:], closestKey)
+	require.Equal(t, closestSum, uint64(9))
+
+	path = sha256.Sum256([]byte("3xwordA250wordB7"))
+	flipPathBit(path[:], 245)
+	closestKey, closestValueHash, closestSum, proof, err = smst.ProveClosest(path[:])
+	require.NoError(t, err)
+	require.NotEqual(t, proof, &SparseMerkleProof{})
+
+	result = VerifySumProof(proof, root, closestKey, closestValueHash, closestSum, NoPrehashSpec(sha256.New(), true))
+	require.True(t, result)
+	closestPath = sha256.Sum256([]byte("3xwordA250wordB7"))
+	require.Equal(t, closestPath[:], closestKey)
+	require.Equal(t, closestSum, uint64(11))
+}
+
+// ProveClosest test against a visual representation of the tree
+// See: https://i.imgur.com/cPJObIy.png
+func TestSMST_ProveClosestFromVisual(t *testing.T) {
+	var smn KVStore
+	var smst *SMST
+	var proof *SparseMerkleProof
+	var result bool
+	var root, closestKey, closestValueHash []byte
+	var closestSum uint64
+	var err error
+
+	smn, err = NewKVStore("")
+	require.NoError(t, err)
+	smst = NewSparseMerkleSumTree(smn, sha256.New(), WithValueHasher(nil))
+
+	// insert random values
+	require.NoError(t, smst.Update([]byte("foo"), []byte("oof"), 3))
+	require.NoError(t, smst.Update([]byte("bar"), []byte("rab"), 6))
+	require.NoError(t, smst.Update([]byte("baz"), []byte("zab"), 9))
+	require.NoError(t, smst.Update([]byte("bin"), []byte("nib"), 12))
+	require.NoError(t, smst.Update([]byte("fiz"), []byte("zif"), 15))
+	require.NoError(t, smst.Update([]byte("fob"), []byte("bof"), 18))
+	require.NoError(t, smst.Update([]byte("testKey"), []byte("testValue"), 21))
+	require.NoError(t, smst.Update([]byte("testKey2"), []byte("testValue2"), 24))
+	require.NoError(t, smst.Update([]byte("testKey3"), []byte("testValue3"), 27))
+	require.NoError(t, smst.Update([]byte("testKey4"), []byte("testValue4"), 30))
+
+	root = smst.Root()
+
+	// testKey2 is the child of an inner node which is the child of an extension node
+	// the extension node has the path bounds of [3, 7] by flipping these bits we force
```

Comment:
`path bounds of [3, 7] by flipping these`

1. I think this is missing some punctuation.
2. You mention the bounds [3, 7] here, and then have two separate bit flips: (3, 6) and (3, 7). Can you just clarify that since the extension has these bounds, we can run tests that flip any bits within those bounds?

Reply:
1. Unsure what you mean?
2. We can flip any bits in the 3-6 range and still get the result of testKey2 as these are in the path bounds (flipping other bits changes the path in a way that results in more than a backstep but we are forcing the proof here to do what we want)
	- so flipping bits: 3,4,5,6 result in testKey2
	- flipping bit 7 results in testKey4

Reply:
Let's update the 4 line comment to this:
```
// `testKey2` is the child of an inner node, which is the child of an extension node.
// The extension node has the path bounds of [3, 7]. This means any bits between
// 3-6 can be flipped, and the resulting path would still traverse through the same 
// extension node and lead to testKey2 - the the closest key. However, flipping bit 7
// will lead to testKey4.
```